pkg/controller/pxc/controller_test.go

```diff
@@ -57,3 +69,450 @@ var _ = Describe("PerconaXtraDB Cluster", Ordered, func() {
+			Name:      ns,
+			Namespace: ns,
+		},
+	}
+
+	BeforeAll(func() {
+		By("Creating the Namespace to perform the tests")
+		err := k8sClient.Create(ctx, namespace)
+		Expect(err).To(Not(HaveOccurred()))
+
+		_, err = envtest.InstallCRDs(cfg, envtest.CRDInstallOptions{
+			Paths: []string{filepath.Join("testdata", "cert-manager.yaml")},
+		})
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	AfterAll(func() {
+		By("Deleting the Namespace to perform the tests")
+		_ = k8sClient.Delete(ctx, namespace)
+	})
+
+	Context("delete-proxysql-pvc finalizer specified", Ordered, func() {
+
+		cr, err := readDefaultCR(crName, ns)
+		cr.Finalizers = append(cr.Finalizers, "delete-proxysql-pvc")
+		cr.Spec.SecretsName = "cluster1-secrets"
+		cr.Spec.HAProxy.Enabled = false
+		cr.Spec.ProxySQL.Enabled = true
+		os.Setenv("DISABLE_TELEMETRY", "true")
+
+		sfsWithOwner := appsv1.StatefulSet{}
+		sfs := statefulset.NewProxy(cr)
+
+		It("should read default cr.yaml", func() {
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		It("Should create PerconaXtraDBCluster", func() {
+			Expect(k8sClient.Create(ctx, cr)).Should(Succeed())
+		})
+
+		It("should reconcile once to create user secret and pvc", func() {
+			_, err := reconciler().Reconcile(ctx, ctrl.Request{NamespacedName: crNamespacedName})
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		It("Should create proxysql sts", func() {
+
+			Expect(k8sClient.Get(ctx, types.NamespacedName{
+				Name:      cr.Name + "-proxysql",
+				Namespace: cr.Namespace,
+			}, &sfsWithOwner)).Should(Succeed())
+		})
+
+		It("Should create secrets", func() {
+			secret := &corev1.Secret{}
+			Expect(k8sClient.Get(ctx, types.NamespacedName{
+				Namespace: cr.Namespace,
+				Name:      cr.Spec.SecretsName,
+			}, secret)).Should(Succeed())
+		})
+		It("should create proxysql PVC", func() {
+			for _, claim := range sfsWithOwner.Spec.VolumeClaimTemplates {
+				for i := 0; i < int(*sfsWithOwner.Spec.Replicas); i++ {
+					pvc := claim.DeepCopy()
+					pvc.Labels = sfs.Labels()
+					pvc.Name = strings.Join([]string{pvc.Name, sfsWithOwner.Name, strconv.Itoa(i)}, "-")
+					pvc.Namespace = ns
+					Expect(k8sClient.Create(ctx, pvc)).Should(Succeed())
+				}
+			}
+		})
+
+		pvcList := corev1.PersistentVolumeClaimList{}
+		It("controller should have proxysql pvc", func() {
+			Eventually(func() bool {
+				err := k8sClient.List(ctx,
+					&pvcList,
+					&client.ListOptions{
+						Namespace: cr.Namespace,
+						LabelSelector: labels.SelectorFromSet(map[string]string{
+							"app.kubernetes.io/component": "proxysql",
+						}),
+					})
+				return err == nil
+			}, time.Second*15, time.Millisecond*250).Should(BeTrue())
+			Expect(len(pvcList.Items)).Should(Equal(3))
+		})
+
+		It("should remove pvc finalizers", func() {
+			for _, pvc := range pvcList.Items {
+				pvcPatch := pvc.DeepCopy()
+				pvcPatch.Finalizers = []string{}
+				err := k8sClient.Patch(ctx, pvcPatch, client.MergeFrom(&pvc))
+				Expect(err).To(Succeed())
+			}
+		})
```

Comment:
```suggestion
```

There is no need to delete pvc finalizers if you are checking pvc deletion by `deletionTimestamp`

Reply:
0787ee2